xwiki-rendering-transformations/xwiki-rendering-transformation-macro/src/main/java/org/xwiki/rendering/internal/transformation/macro/MacroTransformation.java

```diff
@@ -354,4 +358,69 @@ public void setMaxRecursions(int maxRecursions)
     {
         this.maxRecursions = maxRecursions;
     }
+
+    @Override
+    public void prepare(Block block)
+    {
+        Syntax syntax = null;
+
+        // Find the current syntax
+        MetaDataBlock metaDataBlock =
+            block.getFirstBlock(new MetadataBlockMatcher(MetaData.SYNTAX), Axes.ANCESTOR_OR_SELF);
+        if (metaDataBlock != null) {
+            syntax = (Syntax) metaDataBlock.getMetaData().getMetaData(MetaData.SYNTAX);
+        }
+
+        prepare(block, syntax, new HashMap<>());
+    }
+
+    public void prepare(Block block, Syntax syntax, Map<String, Macro<?>> knownMacros)
+    {
+        Syntax currentSyntax = syntax;
+
+        // Check if the syntax changes
+        if (block instanceof MetaDataBlock) {
+            Syntax blockSyntax = (Syntax) ((MetaDataBlock) block).getMetaData().getMetaData(MetaData.SYNTAX);
+            if (blockSyntax != null) {
+                currentSyntax = blockSyntax;
+            }
+        }
+
+        // Prepare the block
+        if (block instanceof MacroBlock) {
+            // Prepare the macro
+            MacroBlock macroBlock = (MacroBlock) block;
+
+            // Try to find a known macros
+            Macro<?> macro = knownMacros.get(macroBlock.getId());
+
+            // If not found use the macro manager
+            try {
+                if (macro == null) {
+                    macro =
+                        MacroTransformation.this.macroManager.getMacro(new MacroId(macroBlock.getId(), currentSyntax));
+
+                    // Cache the found macro for later
+                    knownMacros.put(macroBlock.getId(), macro);
```

Comment:
Isn't it dangerous/incorrect to cache macros by id without syntax but to request them with syntax?

Reply:
I actually partially copy/pasted the logic we have in transform but indeed you are right, it's not accurate as the syntax might change (even if I doubt we have macros associated with syntaxes currently).

To be honest I don't think that local cache really worth it compared to simply always calling MacroManager#getMacro, so I will remove it.